### PR TITLE
User places data - handle recommendations in the data accessor

### PR DESCRIPTION
**Pull request by karenav**

### CODE REVIEW

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 1]
```diff
@@ -18,21 +18,30 @@
...
 import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
 import com.google.appengine.api.datastore.Query.Filter;
 import com.google.appengine.api.datastore.Query.FilterOperator;
 
 public class DataAccessor {
 
   private final DatastoreService datastoreService;
-  @VisibleForTesting
+  private static final String INVALID_USER_MSG = "Invalid user ID";
```
Comment (tbarnahor) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: The message was changed in prevous PRs (that are merged to master) to be "User ID may not be null or empty", so just when you merge this pay attention that it doesn't run it over.

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 2]
```diff
@@ -18,21 +18,30 @@
...
+  @VisibleForTesting // Applies to all the following
   static final String USER_ENTITY_NAME = "User";
+  static final String RECOMMENDATION_ENTITY_KIND = "Recommendation";
+  static final String USER_ID_PROPERTY = "UserId";
+  static final String PLACE_ID_PROPERTY = "PlaceId";
+  static final String CHOSEN_PROPERTY = "WasChosenByUser";
+  static final String TRY_AGAIN_PROPERTY = "WasChosenByUser";
```
Comment (tbarnahor) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: "WasNotChosenByUser"?

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 3]
```diff
@@ -47,30 +56,76 @@ public DataAccessor() {
...
+      recommendationEntity.setProperty(CHOSEN_PROPERTY,
+          feedback.chosenPlace().isPresent() && feedback.chosenPlace().get().equals(place));
+      recommendationEntity.setProperty(TRY_AGAIN_PROPERTY, feedback.userTriedAgain());
+      recommendationEntity.setProperty(TIME_PROPERTY, feedback.feedbackTimeInMillis());
+      datastoreService.put(recommendationEntity);
+    }
+  }
+
+  /**
+  * Approches the database and gets the places that were recommended to the user in the past.
```
Comment (tbarnahor) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: approaches

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 4]
```diff
@@ -47,30 +56,76 @@ public DataAccessor() {
...
   public void registerUser(String userId) {
-    checkArgument(!isNullOrEmpty(userId), "Invalid user ID.");
+    checkArgument(!isNullOrEmpty(userId), INVALID_USER_MSG);
     checkArgument(!isRegistered(userId), "User already registered.");
     Entity userEntity = new Entity(USER_ENTITY_NAME, userId);
     datastoreService.put(userEntity);
   }
+
+  /**
+  * Updates the database with information from the user feedback.
```
Comment (tbarnahor) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: Maybe you can explain that the feedback is for each place recommended to the user. So it is more clear why the entity is of a recommendation kind. It confused me a little bit until I understood that each line in this table is for a single place reccommended to the user.

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 5]
```diff
@@ -47,30 +56,76 @@ public DataAccessor() {
...
+          feedback.chosenPlace().isPresent() && feedback.chosenPlace().get().equals(place));
+      recommendationEntity.setProperty(TRY_AGAIN_PROPERTY, feedback.userTriedAgain());
+      recommendationEntity.setProperty(TIME_PROPERTY, feedback.feedbackTimeInMillis());
+      datastoreService.put(recommendationEntity);
+    }
+  }
+
+  /**
+  * Approches the database and gets the places that were recommended to the user in the past.
+  * It is possible to get only places that the user chose, according to their feedback.
```
Comment (tbarnahor) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: But what if `getOnlyPlacesUserChose` is false?

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java [hunk 6]
```diff
@@ -119,13 +121,129 @@ public void registerUser_nullUserId_throwIllegalArgumentException() {
...
+  }
+
+  @Test
+  public void getUserPlacesHistory_allPlaces_getAllPlaces() {
+    datastoreService.put(createRecomEntityByProperties(
+      USER_ID, PLACE_ID_1, /** chosen */ false, /** tried again */ false));
+    datastoreService.put(createRecomEntityByProperties(USER_ID, PLACE_ID_2, /** chosen */ false,
+        false /** tried again */));
+    // Purposely add the same place IDs again, to make sure we get the places only once.
+    // Once making the familiar place chosen, and once leaving in unchosen.
```
Comment (tbarnahor) [inline project/src/test/java/com/google/sps/data/DataAccessorTest.java]: leaving it unchosen

**Review by tbarnahor — COMMENTED** [review]
Nice PR! Good testing

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 5 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: I think it's clear from the rest of the documentation, you don't think so?

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 1 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 3 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 2 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: done

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java (hunk 6 above)
Comment (karenav) [reply project/src/test/java/com/google/sps/data/DataAccessorTest.java]: done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 4 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: sure, added :)

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 5 above)
Comment (tbarnahor) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: Oh now that I read it again I understood what you meant in this sentence. It sounds like the only possibilty is to get places the user chose, so it didn't go along with the rest of the documentation. Maybe "there is apossibility to get.." will be more clear.

**Review by tbarnahor — APPROVED** [review]

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 5 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: np done

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 7]
```diff
@@ -49,41 +58,89 @@ public DataAccessor() {
...
   public boolean isRegistered(String userId) {
-    checkArgument(!isNullOrEmpty(userId), "User ID may not be null or empty");
+    checkArgument(!isNullOrEmpty(userId), INVALID_USER_MSG);
     Key userIdKey = KeyFactory.createKey(USER_ENTITY_KIND, userId);
     Filter userIdFilter =
         new Query.FilterPredicate(Entity.KEY_RESERVED_PROPERTY, FilterOperator.EQUAL, userIdKey);
     Query query = new Query(USER_ENTITY_KIND).setFilter(userIdFilter).setKeysOnly();
-    return datastoreService.prepare(query)
+    return !datastoreService.prepare(query)
         .asList(FetchOptions.Builder.withDefaults())
```
Comment (szwbar) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: What about using the `countEntities(FetchOptions.withDefaults())` API instead?

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 8]
```diff
@@ -49,41 +58,89 @@ public DataAccessor() {
...
+  *
+  * @param feedback a UserFeedback that holds all the information about the user feedback.
+  */
+  public void updateUserFeedback(UserFeedback feedback) {
+    for (String place : feedback.recommendedPlaces()) {
+      Entity recommendationEntity = new Entity(RECOMMENDATION_ENTITY_KIND);
+      recommendationEntity.setProperty(USER_ID_PROPERTY, feedback.userId());
+      recommendationEntity.setProperty(PLACE_ID_PROPERTY, place);
+      recommendationEntity.setProperty(CHOSEN_PROPERTY,
+          feedback.chosenPlace().isPresent() && feedback.chosenPlace().get().equals(place));
```
Comment (szwbar) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: Optional nit: shorter form may be:
`feedback.chosenPlace().equals(Optional.of(place));`

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 9]
```diff
@@ -49,41 +58,89 @@ public DataAccessor() {
...
+  *                               that the user chose, according to their feedback.
+  * @return the IDs of the places that the user received recommendations about in the past.
+  */
+  public ImmutableList<String> getPlacesRecommendedToUser(String userId,
+      boolean getOnlyPlacesUserChose) {
+    checkArgument(!isNullOrEmpty(userId), INVALID_USER_MSG);
+    Filter userIdFilter =
+        new Query.FilterPredicate(USER_ID_PROPERTY, FilterOperator.EQUAL, userId);
+    Filter chosenPlacesFilter =
+        new Query.FilterPredicate(CHOSEN_PROPERTY, FilterOperator.EQUAL, true);
```
Comment (szwbar) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: Make `chosenPlacesFilter` a static final variable of the class?

File: project/src/main/java/com/google/sps/data/DataAccessor.java [hunk 10]
```diff
@@ -49,41 +58,89 @@ public DataAccessor() {
...
+      recommendationEntity.setProperty(CHOSEN_PROPERTY,
+          feedback.chosenPlace().isPresent() && feedback.chosenPlace().get().equals(place));
+      recommendationEntity.setProperty(TRY_AGAIN_PROPERTY, feedback.userTriedAgain());
+      recommendationEntity.setProperty(TIME_PROPERTY, feedback.feedbackTimeInMillis());
+      datastoreService.put(recommendationEntity);
+    }
+  }
+
+  /**
+  * Approaches the database and gets the places that were recommended to the user in the past.
```
Comment (szwbar) [inline project/src/main/java/com/google/sps/data/DataAccessor.java]: nit: approaches -> queries

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java [hunk 11]
```diff
@@ -123,19 +125,118 @@ public void registerUser_nullUserId_throwIllegalArgumentException() {
...
+    dataAccessor.registerUser(USER_ID);
 
-    assertThrows(IllegalArgumentException.class, () -> dataAccessor.registerUser(userId));
+    assertThrows(IllegalArgumentException.class, () -> dataAccessor.registerUser(USER_ID));
   }
 
   @Test
-  public void storeUserPreferences_validUserIdAndPreferences_userPreferencesStored() {
-    String userId = "12345";
+  public void updateUserFeedback_userChosePlace_UpdateFeedback() {
```
Comment (szwbar) [inline project/src/test/java/com/google/sps/data/DataAccessorTest.java]: nit: update lower case (second one)

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java [hunk 12]
```diff
@@ -162,6 +261,34 @@ public void storeUserPreferences_emptydUserId_throwIllegalArgumentException() {
         () -> dataAccessor.storeUserPreferences("", userPrefs));
   }
 
+  private Entity createRecomEntityByProperties(String userId, String placeId, boolean chosen,
```
Comment (szwbar) [inline project/src/test/java/com/google/sps/data/DataAccessorTest.java]: nit: why not `createRecommendationEntity`?

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java [hunk 13]
```diff
@@ -162,6 +261,34 @@ public void storeUserPreferences_emptydUserId_throwIllegalArgumentException() {
...
+    Entity recommendationEntity = new Entity(DataAccessor.RECOMMENDATION_ENTITY_KIND);
+    recommendationEntity.setProperty(DataAccessor.USER_ID_PROPERTY, userId);
+    recommendationEntity.setProperty(DataAccessor.PLACE_ID_PROPERTY, placeId);
+    recommendationEntity.setProperty(DataAccessor.TIME_PROPERTY, System.currentTimeMillis());
+    recommendationEntity.setProperty(DataAccessor.TRY_AGAIN_PROPERTY, tryAgain);
+    recommendationEntity.setProperty(DataAccessor.CHOSEN_PROPERTY, chosen);
+    return recommendationEntity;
+  }
+
+  private List<Entity> getRecommendationEntitiesByUserId(String userId) {
```
Comment (szwbar) [inline project/src/test/java/com/google/sps/data/DataAccessorTest.java]: replace `get` with `fetch` to make it clearer that this is calling the data store.

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java [hunk 14]
```diff
@@ -162,6 +261,34 @@ public void storeUserPreferences_emptydUserId_throwIllegalArgumentException() {
...
+  }
+
+  private UserFeedback buildUserFeedback(String userId, String chosenPlace,
+      ImmutableList<String> places, boolean userTriedAgain) {
+    UserFeedback.Builder temp = UserFeedback
+        .builder()
+        .setUserId(userId)
+        .setRecommendedPlaces(places)
+        .setUserTriedAgain(userTriedAgain);
+    return chosenPlace == null ? temp.build() : temp.setChosenPlace(chosenPlace).build();
```
Comment (szwbar) [inline project/src/test/java/com/google/sps/data/DataAccessorTest.java]: For readability sake, I think you should separate this to two statements: 

```
if (chosenPlace != null) { temp.setChosenPlace(...); }
return chosenPlace.build();
```

**Review by szwbar — APPROVED** [review]
Great PR! Minor nits only.

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java (hunk 14 above)
Comment (karenav) [reply project/src/test/java/com/google/sps/data/DataAccessorTest.java]: Great tip. done

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java (hunk 13 above)
Comment (karenav) [reply project/src/test/java/com/google/sps/data/DataAccessorTest.java]: Great idea, it's much clearer now. done

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java (hunk 12 above)
Comment (karenav) [reply project/src/test/java/com/google/sps/data/DataAccessorTest.java]: No good reason, changed :)

File: project/src/test/java/com/google/sps/data/DataAccessorTest.java (hunk 11 above)
Comment (karenav) [reply project/src/test/java/com/google/sps/data/DataAccessorTest.java]: done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 10 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 9 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: sure, done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 8 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: Great idea, done

File: project/src/main/java/com/google/sps/data/DataAccessor.java (hunk 7 above)
Comment (karenav) [reply project/src/main/java/com/google/sps/data/DataAccessor.java]: Great idea, I didn't notice that option and it's much clearer. Thanks!

**Pull request merged**